File: realtime/ios/webrtc/Sources/Realtime/SignalingAPI.swift
import Foundation

enum SignalingError: LocalizedError {
    case httpError(Int, String)
    case malformedResponse

    var errorDescription: String? {
        switch self {
        case .httpError(let status, let body):
            "Inworld API error (HTTP \(status)): \(body.prefix(200))"
        case .malformedResponse:
            "Malformed response from Inworld API."
        }
    }
}

struct SignalingAPI {
    var credentials: RealtimeCredentials
    var session: URLSession = .shared

    private struct IceServersResponse: Decodable {
        var iceServers: [IceServer]

        enum CodingKeys: String, CodingKey {
            case iceServers = "ice_servers"
        }
    }

    func fetchIceServers() async throws -> [IceServer] {
        if let preFetched = credentials.preFetchedIceServers {
            return preFetched
        }
        var request = URLRequest(url: credentials.apiBaseURL.appending(path: "v1/realtime/ice-servers"))
        request.setValue(credentials.authorizationHeader, forHTTPHeaderField: "Authorization")
        let data = try await perform(request)
        return try JSONDecoder().decode(IceServersResponse.self, from: data).iceServers
    }

    func postOffer(sdp: String) async throws -> String {
        let url = credentials.callsURL ?? credentials.apiBaseURL.appending(path: "v1/realtime/calls")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(credentials.authorizationHeader, forHTTPHeaderField: "Authorization")
        request.setValue("application/sdp", forHTTPHeaderField: "Content-Type")
        request.httpBody = Data(sdp.utf8)
        let data = try await perform(request)
        guard let answer = String(data: data, encoding: .utf8), !answer.isEmpty else {
            throw SignalingError.malformedResponse
        }
        return answer
    }

    private func perform(_ request: URLRequest) async throws -> Data {
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw SignalingError.malformedResponse }
        guard (200..<300).contains(http.statusCode) else {
            throw SignalingError.httpError(http.statusCode, String(data: data, encoding: .utf8) ?? "")
        }
        return data
    }
}

File: realtime/android/webrtc/app/src/main/java/ai/inworld/voiceagent/realtime/SignalingApi.kt
package ai.inworld.voiceagent.realtime

import ai.inworld.voiceagent.auth.IceServer
import ai.inworld.voiceagent.auth.RealtimeCredentials
import ai.inworld.voiceagent.realtime.events.EventJson
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

class SignalingApi(private val credentials: RealtimeCredentials) {
    @Serializable
    private data class IceServersResponse(
        @SerialName("ice_servers") val iceServers: List<IceServer> = emptyList(),
    )

    suspend fun fetchIceServers(): List<IceServer> {
        credentials.preFetchedIceServers?.let { return it }
        val text = Http.request(
            url = "${credentials.apiBaseUrl}/v1/realtime/ice-servers",
            headers = mapOf("Authorization" to credentials.authorizationHeader),
        )
        return EventJson.decodeFromString<IceServersResponse>(text).iceServers
    }

    suspend fun postOffer(sdp: String): String {
        val url = credentials.callsUrl ?: "${credentials.apiBaseUrl}/v1/realtime/calls"
        val answer = Http.request(
            url = url,
            method = "POST",
            headers = mapOf(
                "Authorization" to credentials.authorizationHeader,
                "Content-Type" to "application/sdp",
            ),
            body = sdp.toByteArray(),
        )
        require(answer.isNotEmpty()) { "Malformed response from Inworld API." }
        return answer
    }
}
